File: benchmarks/fixtures/webapp_rb/middleware/auth_middleware.rb
# Auth middleware (services layer).

require_relative '../utils/helpers'
require_relative '../auth/tokens'
require_relative '../auth/middleware'
require_relative '../exceptions'

# Authentication middleware for the services layer.
class AuthMiddleware
  PUBLIC_PATHS = %w[/health /login /register].freeze

  def initialize(app)
    @app = app
  end

  def call(request)
    validate_request(request)
    return @app.call(request) if PUBLIC_PATHS.include?(request[:path])

    token = extract_token(request)
    raise AuthenticationError.new('Missing token') unless token

    begin
      user = validate_token(token)
      request[:user] = user
      request[:authenticated] = true
      @app.call(request)
    rescue TokenError
      get_logger('middleware.auth').warn('Token validation failed')
      raise AuthenticationError.new('Invalid token')
    end
  end
end

File: benchmarks/fixtures/webapp_php/middleware/auth_middleware.php
<?php

namespace App\Middleware;

use App\AuthenticationError;
use App\Auth\TokenError;

use function App\Auth\extract_token;
use function App\Auth\validate_token;
use function App\Utils\get_logger;
use function App\Utils\validate_request;

/**
 * Authentication middleware for the request pipeline.
 */
class AuthMiddleware
{
    private const PUBLIC_PATHS = ['/health', '/login', '/register'];

    /** @param callable $app */
    public function __construct(private $app)
    {
    }

    /**
     * @param array<string, mixed> $request
     *
     * @return mixed
     */
    public function call(array $request)
    {
        validate_request($request);
        if (in_array($request['path'], self::PUBLIC_PATHS, true)) {
            return ($this->app)($request);
        }

        $token = extract_token($request);
        if ($token === null) {
            throw new AuthenticationError('Missing token');
        }

        try {
            $request['user'] = validate_token($token);
            $request['authenticated'] = true;

            return ($this->app)($request);
        } catch (TokenError) {
            get_logger('middleware.auth')->warn('Token validation failed');
            throw new AuthenticationError('Invalid token');
        }
    }
}
